File: pipeline/screener.py
"""Sector screener: viable, undervalued, US small caps."""
from concurrent.futures import ThreadPoolExecutor

from .net import sec_json

US_STATES = set(
    "AL AK AZ AR CA CO CT DE FL GA HI ID IL IN IA KS KY LA ME MD MA MI MN MS MO MT NE NV NH NJ NM NY NC ND OH "
    "OK OR PA RI SC SD TN TX UT VT VA WA WV WI WY DC".split()
)

RULES = {
    "max_ps": 2.0,
    "max_mcap": 3e9,
    "bonus_mcap": 2e9,
    "max_lt_de": 0.5,
    "min_net_margin": 0.08,
    "min_profitable_years": 2,
    "min_revenue_growth": -0.05,
    "out_of_favor_drawdown": 0.25,
    "undiscovered_max_analysts": 2,
}


def registration(cik):
    """State of incorporation and business address from the SEC's company record."""
    try:
        s = sec_json(f"https://data.sec.gov/submissions/CIK{cik:010d}.json")
    except Exception:
        return None
    biz = (s.get("addresses") or {}).get("business") or {}
    return {
        "incorporated": (s.get("stateOfIncorporation") or "").upper(),
        "hq": (biz.get("stateOrCountry") or "").upper(),
        "sic": s.get("sicDescription"),
    }


def _checks(u, m):
    """Each hard rule as (key, passed, plain-English note). Missing data counts as a fail."""
    ps = u["mcap"] / m["revenue"] if u.get("mcap") and m.get("revenue") and m["revenue"] > 0 else None
    out = [
        ("ps", ps is not None and ps < RULES["max_ps"], ps),
        ("mcap", bool(u.get("mcap")) and u["mcap"] < RULES["max_mcap"], u.get("mcap")),
        ("debt", m.get("lt_debt_to_equity") is not None and m["lt_debt_to_equity"] < RULES["max_lt_de"], m.get("lt_debt_to_equity")),
        ("margin", m.get("net_margin") is not None and m["net_margin"] >= RULES["min_net_margin"], m.get("net_margin")),
        ("equity", bool(m.get("equity")) and m["equity"] > 0, m.get("equity")),
        ("fcf", m.get("fcf") is not None and m["fcf"] > 0, m.get("fcf")),
        ("profitable", m.get("years_checked", 0) >= 2 and m["profitable_years"] >= RULES["min_profitable_years"], m.get("profitable_years")),
        ("growth", m.get("revenue_growth") is not None and m["revenue_growth"] >= RULES["min_revenue_growth"], m.get("revenue_growth")),
    ]
    return out, ps


def _score(row):
    """0-100 rank score. Every component is listed so the page can explain it."""
    parts = []
    ps = row["ps"]
    parts.append(("Cheap vs. sales", round(max(0.0, (RULES["max_ps"] - ps) / RULES["max_ps"]) * 20, 1), 20))
    if row["total_debt"] <= 0:
        parts.append(("No debt at all", 20.0, 20))
    elif row["net_cash"] > 0:
        parts.append(("More cash than debt", 14.0, 20))
    else:
        parts.append(("Low debt", round(max(0.0, 1 - row["lt_de"] / RULES["max_lt_de"]) * 8, 1), 20))
    parts.append(("Profit margin", round(min(row["net_margin"], 0.30) / 0.30 * 15, 1), 15))
    fy = row["fcf_yield"] or 0
    parts.append(("Cash generation", round(max(0.0, min(fy, 0.15)) / 0.15 * 15, 1), 15))
    g = row["revenue_growth"] or 0
    parts.append(("Sales growth", round(max(0.0, min(g, 0.20)) / 0.20 * 10, 1), 10))
    parts.append(("Under $2B", 5.0 if row["mcap"] < RULES["bonus_mcap"] else 0.0, 5))
    dd = row["drawdown"]
    parts.append(("Out of favor", round(max(0.0, min(dd or 0, 0.5)) / 0.5 * 8, 1), 8))
    a = row["analysts"]
    parts.append(("Few analysts", 7.0 if a == 0 else 4.0 if a is not None and a <= RULES["undiscovered_max_analysts"] else 0.0, 7))
    return round(sum(p[1] for p in parts)), [{"label": l, "points": p, "max": mx} for l, p, mx in parts]


def run(universe, metrics, prices, analyst_counter):
    """universe/metrics/prices keyed by symbol. Returns the screener JSON payload."""
    candidates, checked_by_sector = [], {}
    for sym, u in universe.items():
        m = metrics.get(sym)
        checked_by_sector[u["sector"]] = checked_by_sector.get(u["sector"], 0) + 1
        if not m or u["country"] not in ("United States", ""):
            continue
        if not (m.get("loc") or "").startswith("US"):
            continue
        checks, ps = _checks(u, m)
        if all(ok for _, ok, _ in checks):
            candidates.append((sym, ps))

    # Only survivors need the slower per-company lookups.
    syms = [s for s, _ in candidates]
    with ThreadPoolExecutor(max_workers=4) as pool:
        regs = dict(zip(syms, pool.map(lambda s: registration(universe[s]["cik"]), syms)))
    us = [s for s in syms if regs.get(s) and regs[s]["incorporated"] in US_STATES and regs[s]["hq"] in US_STATES]
    analysts = analyst_counter.counts(us)

    rows = []
    ps_by = dict(candidates)
    for sym in us:
        u, m, p = universe[sym], metrics[sym], prices.get(sym) or {}
        price = p.get("price") or u["price"]
        high = p.get("high52")
        drawdown = (1 - price / high) if high else None
        row = {
            "symbol": sym,
            "name": u["name"],
            "sector": u["sector"],
            "industry": u["industry"],
            "hq": regs[sym]["hq"],
            "price": price,
            "mcap": u["mcap"],
            "ps": round(ps_by[sym], 3),
            "net_margin": m["net_margin"],
            "lt_de": m["lt_debt_to_equity"],
            "total_debt": m["total_debt"],
            "net_cash": m["net_cash"],
            "fcf": m["fcf"],
            "fcf_yield": (m["fcf"] / u["mcap"]) if u["mcap"] else None,
            "revenue": m["revenue"],
            "revenue_growth": m["revenue_growth"],
            "profitable_years": m["profitable_years"],
            "years_checked": m["years_checked"],
            "fiscal_year": m["fiscal_year"],
            "high52": high,
            "drawdown": drawdown,
            "analysts": analysts.get(sym),
        }
        row["debt_status"] = "none" if row["total_debt"] <= 0 else "net_cash" if row["net_cash"] > 0 else "low"
        row["flags"] = []
        if drawdown is not None and drawdown >= RULES["out_of_favor_drawdown"]:
            row["flags"].append("out_of_favor")
        if row["analysts"] is not None and row["analysts"] <= RULES["undiscovered_max_analysts"]:
            row["flags"].append("undiscovered")
        row["score"], row["score_parts"] = _score(row)
        rows.append(row)

    rows.sort(key=lambda r: -r["score"])
    sectors = sorted({u["sector"] for u in universe.values()})
    return {
        "rules": RULES,
        "sectors": [
            {"name": s, "checked": checked_by_sector.get(s, 0), "passed": sum(1 for r in rows if r["sector"] == s)}
            for s in sectors
        ],
        "results": rows,
    }

File: site/assets/screener.js
import { chrome, esc, load, meter, money, pct, price, times } from "./common.js";

chrome();

const state = { sector: new URLSearchParams(location.search).get("sector") || "All", sort: "score", open: null };
const host = document.querySelector("[data-results]");
const countEl = document.querySelector("[data-count]");
let data;

const SORTS = {
  score: (a, b) => b.score - a.score,
  ps: (a, b) => a.ps - b.ps,
  net_margin: (a, b) => b.net_margin - a.net_margin,
  fcf_yield: (a, b) => (b.fcf_yield ?? -1) - (a.fcf_yield ?? -1),
  drawdown: (a, b) => (b.drawdown ?? -1) - (a.drawdown ?? -1),
  mcap: (a, b) => a.mcap - b.mcap,
};

const band = (s) => (s >= 70 ? "strong" : s >= 40 ? "fair" : "weak");
const debtTag = (r) =>
  r.debt_status === "none" ? `<span class="tag tag-up">No debt</span>`
  : r.debt_status === "net_cash" ? `<span class="tag tag-up">Net cash</span>`
  : `<span class="tag">Low, ${times(r.lt_de)}</span>`;

function flags(r) {
  const f = [];
  if (r.flags.includes("out_of_favor")) f.push(`<span class="tag tag-mid" title="25% or more below its 52-week high">Out of favor</span>`);
  if (r.flags.includes("undiscovered")) f.push(`<span class="tag" title="Two or fewer analysts cover it">Undiscovered</span>`);
  return f.length ? `<div class="co-flags">${f.join("")}</div>` : "";
}

function facts(r) {
  const out = [];
  out.push(["ph-receipt", `Sold ${money(r.revenue)} in fiscal ${r.fiscal_year} and kept ${pct(r.net_margin, 0)} of it as profit${r.one_time_note ? ", not counting one-time items" : ""}.`]);
  if (r.one_time_note) out.push(["ph-info", r.one_time_note]);
  if (r.revenue_growth != null) out.push(["ph-trend-up", `Sales ${r.revenue_growth >= 0 ? "grew" : "slipped"} ${pct(Math.abs(r.revenue_growth))} from the year before.`]);
  out.push(["ph-coins", r.debt_status === "none" ? `Has no debt. Cash on hand: ${money(r.net_cash)}.`
    : r.net_cash > 0 ? `Holds ${money(r.net_cash)} more cash than all of its debt.`
    : `Long-term debt is ${times(r.lt_de)} its equity, under the 0.5 limit.`]);
  if (r.fcf_yield != null) out.push(["ph-hand-coins", `Generated ${money(r.fcf)} of free cash, ${pct(r.fcf_yield)} of its market value.`]);
  const asReported = r.profitable_years_reported;
  out.push(["ph-calendar-check", `Profitable in ${r.profitable_years} of the last ${r.years_checked} years${
    asReported != null && asReported !== r.profitable_years ? `, not counting one-time items (${asReported} as reported)` : ""}.`]);
  if (r.drawdown != null) out.push(["ph-arrow-down-right", `Trades ${pct(r.drawdown, 0)} below its 52-week high of ${price(r.high52)}.`]);
  out.push(["ph-users", r.analysts == null ? "Analyst coverage could not be checked today." : r.analysts === 0 ? "No Wall Street analysts publish estimates on it." : `${r.analysts} analyst${r.analysts === 1 ? "" : "s"} publish estimates on it.`]);
  return out;
}

function detailRow(r, cols) {
  // Whole points when the pipeline sends them. Older data has one-decimal parts, and rounding those
  // one by one here would stop them adding up to the score, so they keep their decimal.
  const whole = r.score_parts.every((p) => Number.isInteger(p.points));
  const pts = (n) => (whole ? String(n) : (Math.round(n * 10) / 10).toFixed(1));
  const total = r.score_parts.reduce((a, p) => a + p.points, 0);
  return `<tr class="detail"><td colspan="${cols}"><div class="detail-inner">
    <div><h4>Why it passed</h4><ul class="facts">${facts(r).map(([i, t]) => `<li><i class="ph ${i}" aria-hidden="true"></i><span>${esc(t)}</span></li>`).join("")}</ul></div>
    <div><h4>How the ${r.score} score adds up</h4><div class="parts">${r.score_parts.map((p) => `
      <div class="part"><span>${esc(p.label)}</span><div class="part-track"><span style="width:${(p.points / p.max) * 100}%"></span></div><span class="pts">${pts(p.points)} / ${p.max}</span></div>`).join("")}
      <div class="part part-total"><span>Total</span><span>${Math.abs(total - r.score) < 0.05 ? "" : `Rounds to ${r.score}`}</span><span class="pts">${pts(total)} / ${r.score_parts.reduce((a, p) => a + p.max, 0)}</span></div>
    </div></div>
    <div class="detail-actions"><a class="btn btn-primary btn-sm" href="report.html?t=${encodeURIComponent(r.symbol)}">Full report on ${esc(r.symbol)}</a></div>
  </div></td></tr>`;
}

function renderSectors() {
  const el = document.querySelector("[data-sectors]");
  const total = data.results.length;
  const list = [{ name: "All", passed: total }, ...data.sectors.filter((s) => s.checked > 0 || s.excluded)];
  el.innerHTML = list.map((s) => `<button class="pill" type="button" data-sector="${esc(s.name)}" aria-pressed="${s.name === state.sector}">
    ${esc(s.name === "All" ? "All sectors" : s.name)}<span class="count">${s.passed}</span></button>`).join("");
  el.onclick = (e) => {
    const b = e.target.closest("[data-sector]");
    if (!b) return;
    state.sector = b.dataset.sector;
    state.open = null;
    const u = new URL(location);
    state.sector === "All" ? u.searchParams.delete("sector") : u.searchParams.set("sector", state.sector);
    history.replaceState(null, "", u);
    renderSectors();
    renderTable();
  };
}

// Companies that passed every other check but were left out because a check could not be confirmed on this run:
// their US registration ("us") or their debt ("debt"), each listed in "unconfirmed" with its reason in "reason".
// Older data lists only registration problems, as symbols or as objects without "unconfirmed".
const its = (n) => (n === 1 ? "its" : "their");
const isLeftOut = (n) => (n === 1 ? "it is left out" : "they are left out");
const UNCONFIRMED = {
  us: (n) => `passed every financial check, but ${its(n)} US headquarters or incorporation could not be confirmed with the SEC today, so ${isLeftOut(n)}`,
  debt: (n) => `passed every other check, but ${its(n)} debt could not be read reliably from ${its(n)} filings, so the low-debt rule could not be confirmed and ${isLeftOut(n)}`,
  "debt us": (n) => `passed every other check, but ${its(n)} debt could not be read reliably from ${its(n)} filings and ${its(n)} US headquarters or incorporation could not be confirmed with the SEC today, so ${isLeftOut(n)}`,
};
const andList = (a) => (a.length < 2 ? a.join("") : `${a.slice(0, -1).join(", ")} and ${a[a.length - 1]}`);

function unverifiedNote() {
  const list = (data.unverified || []).map((u) => (typeof u === "string" ? { symbol: u } : u))
    .filter((u) => u && u.symbol && (state.sector === "All" || !u.sector || u.sector === state.sector));
  if (!list.length) return "";
  const groups = new Map();
  for (const u of list) {
    const key = Array.isArray(u.unconfirmed) && u.unconfirmed.length ? [...u.unconfirmed].sort().join(" ") : "us";
    if (!groups.has(key)) groups.set(key, []);
    groups.get(key).push(u);
  }
  const order = ["us", "debt", "debt us"];
  const keys = [...groups.keys()].sort((a, b) => (order.indexOf(a) + 1 || 9) - (order.indexOf(b) + 1 || 9));
  const sentences = keys.map((key) => {
    const g = groups.get(key);
    const n = g.length;
    const links = andList(g.map((u) => {
      const title = [u.name && u.name.replace(/\.$/, ""), u.reason].filter(Boolean).join(". ");
      return `<a href="report.html?t=${encodeURIComponent(u.symbol)}"${title ? ` title="${esc(title)}"` : ""}>${esc(u.symbol)}</a>`;
    }));
    const why = UNCONFIRMED[key] ? UNCONFIRMED[key](n)
      : `passed every other check, but one check could not be confirmed today, so ${isLeftOut(n)}`;
    // A registration lookup that failed today may work tomorrow. A debt reading changes only with new filings.
    const again = key.split(" ").includes("us") ? " The next update checks again." : "";
    return `${n} ${n === 1 ? "company" : "companies"} ${why}: ${links}.${again}`;
  });
  return `<div class="notice unverified"><i class="ph ph-info" aria-hidden="true"></i>
    <div>${sentences.map((s, i) => `<p${i ? ` style="margin-top:6px"` : ""}>${s}</p>`).join("")}</div></div>`;
}

function renderTable() {
  const rows = data.results.filter((r) => state.sector === "All" || r.sector === state.sector).sort(SORTS[state.sort]);
  const sec = data.sectors.find((s) => s.name === state.sector);
  countEl.innerHTML = state.sector === "All"
    ? `<b>${rows.length}</b> ${rows.length === 1 ? "company" : "companies"} passed every check`
    : sec?.excluded ? `${esc(state.sector)} is not screened`
    : `<b>${rows.length}</b> of ${sec?.checked ?? 0} ${esc(state.sector)} companies passed every check`;

  if (sec?.excluded) {
    host.innerHTML = `<div class="panel empty"><h3>${esc(state.sector)} companies are not screened</h3><p>${esc(sec.excluded)}</p></div>`;
    return;
  }

  if (!rows.length) {
    host.innerHTML = `<div class="panel empty"><h3>No ${esc(state.sector)} company passes today</h3>
      <p>That is normal. The checks are strict on purpose, and some sectors rarely have cheap, debt-light, profitable small caps.
      Try another sector or come back after the next update.</p></div>${unverifiedNote()}`;
    return;
  }
  const cols = 10;
  host.innerHTML = `<div class="table-wrap"><table class="data">
    <thead><tr>
      <th>#</th><th>Company</th><th style="text-align:left">Score</th><th>Market value</th><th>Price / sales</th>
      <th>Net margin</th><th>Debt</th><th>Cash yield</th><th>Below high</th><th>Analysts</th>
    </tr></thead>
    <tbody>${rows.map((r, i) => `
      <tr class="row" tabindex="0" data-sym="${esc(r.symbol)}" aria-expanded="${state.open === r.symbol}">
        <td class="rank-cell">${i + 1}</td>
        <td class="co-cell"><div class="co"><span class="sym">${esc(r.symbol)}</span><span class="nm" title="${esc(r.name)}">${esc(r.name)}</span></div>${flags(r)}</td>
        <td class="score-cell">${meter(r.score, band(r.score))}</td>
        <td>${money(r.mcap)}</td>
        <td>${times(r.ps)}</td>
        <td${r.one_time_note && r.net_margin_reported != null ? ` title="Leaves out one-time items. Reported margin: ${pct(r.net_margin_reported)}"` : ""}>${pct(r.net_margin)}</td>
        <td>${debtTag(r)}</td>
        <td>${pct(r.fcf_yield)}</td>
        <td>${r.drawdown == null ? "n/a" : pct(r.drawdown, 0)}</td>
        <td>${r.analysts ?? "n/a"}</td>
      </tr>${state.open === r.symbol ? detailRow(r, cols) : ""}`).join("")}
    </tbody></table></div>
    <p class="faint" style="font-size:13px;margin-top:12px">Select a row to see why it passed and how its score adds up. Cash yield is free cash flow as a share of market value.</p>
    ${unverifiedNote()}`;

  host.querySelectorAll("tr.row").forEach((tr) => {
    const toggle = () => { state.open = state.open === tr.dataset.sym ? null : tr.dataset.sym; renderTable(); };
    tr.addEventListener("click", toggle);
    tr.addEventListener("keydown", (e) => { if (e.key === "Enter" || e.key === " ") { e.preventDefault(); toggle(); } });
  });
}

document.querySelector("[data-sort]").addEventListener("change", (e) => { state.sort = e.target.value; renderTable(); });

load("screener.json").then((d) => {
  data = d;
  if (state.sector !== "All" && !d.sectors.some((s) => s.name === state.sector)) state.sector = "All";
  renderSectors();
  renderTable();
}).catch(() => {
  host.innerHTML = `<div class="panel empty"><h3>Screener data is not available yet</h3><p>Results appear after the first nightly update runs.</p></div>`;
});
